samples/guide/src/main/java/okhttp3/recipes/kt/WiresharkExample.kt

```diff
@@ -0,0 +1,287 @@
+  }
+
+  override fun secureConnectEnd(
+    call: Call,
+    handshake: Handshake?
+  ) {
+    logger.removeHandler(loggerHandler)
+  }
+
+  override fun callEnd(call: Call) {
+    // Cleanup log handler if failed.
+    logger.removeHandler(loggerHandler)
+  }
+
+  override fun connectionAcquired(
+    call: Call,
+    connection: Connection
+  ) {
+    if (random != null) {
+      val sslSocket = connection.socket() as SSLSocket
+      val session = sslSocket.session
+
+      val masterSecretHex = session.masterSecret?.encoded?.toByteString()
+          ?.hex()
+
+      if (masterSecretHex != null) {
+        val keyLog = "CLIENT_RANDOM $random $masterSecretHex"
+
+        if (verbose) {
+          println(keyLog)
+        }
+        logFile.appendText("$keyLog\n")
+      }
+    }
+
+    random = null
+  }
+
+  class Factory(
+    private val logFile: File,
+    private val verbose: Boolean = false
+  ) : EventListener.Factory {
+    override fun create(call: Call): EventListener {
+      return WireSharkKeyLoggerListener(logFile, verbose)
+    }
+  }
+
+  companion object {
+    private lateinit var logger: Logger
+
+    private val SSLSession.masterSecret: SecretKey?
+      get() = javaClass.getDeclaredField("masterSecret")
+          .apply {
+            isAccessible = true
+          }
+          .get(this) as? SecretKey
+
+    val randomRegex = "\"random\"\\s+:\\s+\"([^\"]+)\"".toRegex()
+
+    fun register() {
+      // Enable JUL logging for SSL events, must be activated early or via -D option.
+      System.setProperty("javax.net.debug", "")
+      logger = Logger.getLogger("javax.net.ssl")
+          .apply {
+            level = Level.FINEST
+          }
+    }
+  }
+}
+
+val eventListenerFactory = WireSharkKeyLoggerListener.Factory(File("/tmp/key.log"), verbose = true)
+
+@SuppressSignatureCheck
+class WiresharkExample(private val tlsVersions: List<TlsVersion>) {
+  val connectionSpec =
+    Builder(ConnectionSpec.RESTRICTED_TLS)
+        .tlsVersions(*tlsVersions.toTypedArray())
+        .build()
+
+  var bootstrapClient = OkHttpClient.Builder()
+      .connectionSpecs(listOf(connectionSpec))
+      .addInterceptor(BrotliInterceptor)
+      .eventListenerFactory(eventListenerFactory)
+      .build()
+
+  val dns = DnsOverHttps.Builder()
+      .client(bootstrapClient)
+      .url("https://1.1.1.1/dns-query".toHttpUrl())
+      .includeIPv6(false)
+      .build()
+
+  val client = OkHttpClient.Builder()
+      .connectionSpecs(listOf(connectionSpec))
+      .eventListenerFactory(eventListenerFactory)
+      .addInterceptor(BrotliInterceptor)
+      .dns(dns)
+      .build()
+
+  fun run() {
+    if (tlsVersions.contains(TLS_1_2)) {
+      println("TLSv1.2 traffic will be logged automatically")
+    }
+
+    if (tlsVersions.contains(TLS_1_3)) {
+      println("TLSv1.3 requires an external command run before first traffic is sent")
+      println("Follow instructions at https://github.com/neykov/extract-tls-secrets for TLSv1.3")
```

Comment:
The agent is quite interesting. I wonder if we could detect it automatically so you don’t need to do all the work to extract the PID and log

Reply:
The agent is awesome - nice work @neykov

If you are TLSv1.3 - the agent is as automatic as you need since you can just run remotely against any running Java app.  OkHttp or otherwise.

But this sample is mainly to make it simple without running additional steps internally.  Not sure how to incrementally improve for now.  Let's try this out ourselves debugging failures against QNAP or Go based servers :)